### PR TITLE
#86 - Adds enclosing ranges to function and class definitions

### PR DESCRIPTION
**Pull request by Titou325**
This is a first attempt at implementing [enclosing ranges](https://github.com/sourcegraph/scip/blame/2314da70dd146097c09ca3ec1326c5bede5e7af6/scip.proto#L624-L663) for function and class definitions, closing #86.

I have attempted to keep it as simple as possible, passing a parent node if needed to [`pushNewOccurrence`](https://github.com/Titou325/scip-python/blob/e0a3d30a753fbbe3077d77b74cbd3b01a4ed7a47/packages/pyright-scip/src/treeVisitor.ts#L1397).

This does not implement enclosing ranges for reference occurences.

These are the pending items:
- [x] Add additional test cases covering specifics (decorators, type hints, stubs) @Titou325
- [x] Check for possible regressions due to updating `scip.proto`
- [x] Ensure functionality works as designed (ping @varungandhi-src) for the Python language
- [x] add `# format-options: showDocs` to other test snapshot inputs and regenerate expected outputs  (currently blocked by #91)
- [ ] Update documentation?

Any feedback is very welcome. Until we can update the remaining snapshots at the very least, this PR will stay in draft.
Thanks!

### CODE REVIEW

**Comment by jdorfman** [comment]
Thanks @Titou325 

@varungandhi-src go any feedback?

File: packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py
Comment (varungandhi-src) [inline packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py]: The point of this test case is to test the enclosing ranges.

1. The output around documentation is distracting from the result. In scip-clang, we support magic comments (https://sourcegraph.com/search?q=context:global+repo:%5Egithub%5C.com/sourcegraph/scip-clang%24%40main+format-options:&patternType=standard&sm=1&groupBy=path) to customize what is printed in the snapshot output. I would recommend adding support for that, so that we don't have doc comments distracting from what's actually under test.
2. If someone adds a test case in-between the existing test cases, then the numbers shown in the enclosing ranges for unrelated declarations will change. This is one of the reasons why the snapshots use `^^^` markers instead of actual numbers. I suggest changing the output style to use `^ start enclosing_range <symbol_name>` and `^ end enclosing_range <symbol_name>` instead. Cross-checking whether the line and character numbers are correct also requires looking at two files at once, the original and the snapshot file. The usage of carets would avoid that.

**Review by varungandhi-src — CHANGES_REQUESTED** [review]
I haven't had the time to look at this fully, but I recommend first changing the snapshot output format to be something more digestible.

File: packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py
Comment (Titou325) [reply packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py]: Thanks for the feedback, @varungandhi-src 

Two things done:
- [x] added support for magic comments in the form of `# format-options: showDocs, showRanges`
- [x] switched to inline markers in the form of `⌄ start enclosing_range` and `⌃ end enclosing_range`.

**Comment by Titou325** [comment]
Hi @varungandhi-src,

Thanks for the feedback. I have updated the snapshot format to match your expectations, with the difference of the start enclosing range being a down-pointing caret. It seems to me that this would be most natural given that the start range information is given above the first line (due to possible decorators presence), while the end range is below the last line of the range.

This PR is ready for review but will require subsequent modifications to merge into main, although they are currently blocked by #91.

File: packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py
Comment (varungandhi-src) [reply packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py]: Thanks for the update, I'll take a deeper look in the next 24 hours.

File: packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py
Comment (varungandhi-src) [inline packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py]: nit: This formatting is a bit off. Note that the definition line above uses a special `<` to indicate that the range begins at the start of the line, we should use similar logic for `begin` and `end` here.

Comment (Titou325) [reply packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py]: Agreed, although said syntax makes quick overview a bit harder. I would be happy to implement a standard to snapshots across scip-* repositories.

Comment (varungandhi-src) [reply packages/pyright-scip/snapshots/output/enclosing_range_class/advanced.py]: > Agreed, although said syntax makes quick overview a bit harder

It's important that we have:
1. Consistent syntax for different source range annotations
2. Be able to mark the start of the line/end of the line.

Another alternative is to just offset everything by +2, which is done in some other repos, but in Python's case, it makes browsing the snapshot files more annoying in an editor because of the red squiggles.

That said, I'm not married to the exact syntax being used (if you have other suggestions, I'm all ears). The important criteria are 1 and 2 above.

---

> I would be happy to implement a standard to snapshots across scip-* repositories.

While we do have some core functionality in `scip snapshot` here: https://github.com/sourcegraph/scip/blob/main/docs/CLI.md#scip-snapshot

We don't use it in the other indexer repositories because:
1. The various indexers are in different languages, so we can't reuse the snapshotting logic as a library easily (it is in Go).
4. It's valuable to quickly be able to customize the snapshotting logic depending on the use-case without having to change it in a central place.

It's not that much code overall, about 100-200 lines, but having it be different allows moving different repos in a more de-coupled way.

File: packages/pyright-scip/src/treeVisitor.ts
Comment (varungandhi-src) [inline packages/pyright-scip/src/treeVisitor.ts]:
```suggestion
        // TODO(issue: https://github.com/sourcegraph/scip-python/issues/134)
        // Support more kinds of nodes here
```

File: packages/pyright-scip/src/lib.ts
Comment (varungandhi-src) [inline packages/pyright-scip/src/lib.ts]: This function is already getting quite long, it would be easier to read if this logic were hoisted out into a separate function that returns a `{ showDocs: boolean, showRanges: boolean}` object.

File: packages/pyright-scip/src/lib.ts
Comment (varungandhi-src) [inline packages/pyright-scip/src/lib.ts]: This data will be thrown away later if `!formatDocs.showRanges`. So instead of the early exit later on, it would be simpler to change this condition to be 
```suggestion
        if (formatOptions.showRanges && occurrence.enclosing_range.length > 0) {
            enclosingRanges.push({
                range: Range.fromLsif(occurrence.enclosing_range),
                symbol: occurrence.symbol,
            });
        }
```
and remove the 

```
        if (!formatOptions.showRanges) {
            return;
        }
```

later in the code.

File: packages/pyright-scip/src/lib.ts
Comment (varungandhi-src) [inline packages/pyright-scip/src/lib.ts]: I think the logic here and below would be simplified quite a bit if instead of the sort earlier + lookup here, you instead did something like creating two lists outside this loop:

```
const enclosingRangeStarts = Array.from(Array(inputs.lines.length), () => [])
const enclosingRangeEnds = Array.from(Array(inputs.lines.length), () => [])
for (const enclosingRange of enclosingRanges) {
  // Push entry to start, and ends array based on line numbers
}
```

Later, when iterating over the lines, you can sort the sub-arrays at index `i` based on the line, and consume that, instead of having to do a lookup through the list of ranges. The sort order would take care of the iteration, and avoid the need for element removal or +1/-1 logic.

File: packages/pyright-scip/src/treeVisitor.ts
Comment (varungandhi-src) [inline packages/pyright-scip/src/treeVisitor.ts]:
```suggestion
```

The code below is clear that we're checking for the function case. 😄

File: packages/pyright-scip/src/treeVisitor.ts
Comment (varungandhi-src) [inline packages/pyright-scip/src/treeVisitor.ts]: Not exactly part of your patch, but I think this code would be clearer if the usages of `parent` in this branch would be replaced by `decl.node`. IIUC, that is going to be true, because we're in the `isDefinition = decl.node.is == parent.id` case, but it's a little confusing to read this because there are seemingly 3 different things here (`node`, `decl.node`, `parent`):

```
                    this.pushNewOccurrence(node, this.getScipSymbol(decl.node), scip.SymbolRole.Definition, parent);
```

whereas the code path earlier for enclosing_ranges has:

```
this.pushNewOccurrence(node, existingSymbol, scip.SymbolRole.Definition, decl.node);
```

where `existingSymbol` was derived from `decl.node` just a few lines earlier.

**Comment by varungandhi-src** [comment]
> This PR is ready for review but will require subsequent modifications to merge into main, although they are currently blocked by https://github.com/sourcegraph/scip-python/issues/91

We don't need to be blocked by that issue. Either I or you can run the snapshots on Ubuntu and push that once the code changes are done.

**Review by varungandhi-src — APPROVED** [review]
Overall, I think this a good improvement (new feature + better testing), so thank you for the work!

I've left some style-related comments which I think would help make the code easier to follow. Most of the them are non-blocking, but it'd be nice to at least simplify the enclosing ranges logic in the snapshot code to some extent.

Tentatively approving. Let me know if you'd like to update the snapshots on Ubuntu and push (or you'd like me to do that).

**Comment by varungandhi-src** [comment]
> Check for possible regressions due to updating scip.proto

`scip.proto` evolves in a backwards compatible way, so it shouldn't affect indexer output.

**Comment by Titou325** [comment]
Hi!

Updated the PR regarding all points above mentioned. I don't have a Linux runner easily accessible at this time, would be glad if you could update the snapshots @varungandhi-src. N.B. the `showDocs` option will probably be required on the other files.

Thanks a lot for the feedback!

**Comment by varungandhi-src** [comment]
@Titou325 I've updated the snapshots here using a Linux runner: https://github.com/codemuse-app/scip-python/pull/1

**Comment by Titou325** [comment]
Hi @varungandhi-src, I have merged and updated the PR, fixing the issue with duplicate class declarations. This is an org owned fork, so I am unfortunately unable to give blanket permissions but I have given you write access.

The snapshots should be updated on Linux runners and then I believe we are good to go.
Thanks!

**Comment by varungandhi-src** [comment]
@Titou325 if you'd like, I think you should be able to remove my maintain access on your fork, and use this option for future PRs.

**Comment by Titou325** [comment]
Perfect, thanks! Can we bump versions to trigger a new release?

**Comment by maks-ivanov** [comment]
Thanks @Titou325 awesome work

**Pull request merged**
